### ISSUE TITLE
[❔ other question] Unable to Optimize Vertex Positions

### ISSUE THREAD

**Issue by uguuuuuu** [mitsuba-renderer/mitsuba3#169]
## Summary

I wanted to optimize the vertex positions of a mesh but encountered some strange error. 

## System configuration

*For bug report, please enter information regarding your system configuration<- [remove this]*

- Platform: Ubuntu 22.04
- Compiler: clang 14
- Python version: 3.9
- Mitsuba 3 version: 3.0.1
- Compiled variants:
    * `scalar_rgb`
    * `cuda_ad_rgb`

## Description

I called `dr.enable_grad()` on the vertex positions and then `params.update()` to update the scene, and Dr. JIT complained that an unknown variable was referenced. But the error did not occur when the parameter was vertex normals.

## Steps to reproduce

```
import drjit as dr
import mitsuba as mi

mi.set_variant('cuda_ad_rgb')

key = 'redwall.vertex_positions'

# Loading the Cornell Box scene shipped along with Mitsuba 3
scene = mi.load_file('tutorials/scenes/cbox.xml', res=128, integrator='prb_reparam')

params = mi.traverse(scene)

opt = mi.ad.Adam(lr=0.02)
opt[key] = params[key]
params.update(opt)
```
```
Critical failure in Dr.Jit AD backend: referenced an unknown variable a2494425056!
Aborted (core dumped)
```

**Comment by uguuuuuu** [comment]
It seems to have something to do with the compiler. The above code ran fine with Mitsuba compiled using Clang 11.

**Comment by Speierers** [comment]
Can you confirm that this is a problem with Clang 14? In which case I will give it a shot.

**Comment by uguuuuuu** [comment]
> 

I don't have a Ubuntu 22.04 machine now so I can't confirm it now. But I'm sure that both the llvm and cuda variants produced the error when compiled using Clang 14 and everything was fine after I switched to Clang 11, and there was nothing else changed.

**Comment by LiuLinyun** [comment]
@uguuuuuu @njroussel 

I have got this same error message in psdr-cuda (which is another differentiable render powered by enoki (a project successed by Dr.jit) and I solved it.

In my case, I found it is an error caused by the global variable `static State state` shared between different dynamic libraries (.so or .dll) but constructed more than once in runtime.

https://github.com/mitsuba-renderer/drjit/blob/07e9da811e7284b87fa292472b30ec4465592eef/src/autodiff/autodiff.cpp#L546

Compiler compiled one instance in pybind11 based libraries and construct when python importing drjit and another instance is constructed when mitsuba3 linking and loading c++ dynamic libraries without pybind11.
When code is running python put a variable index into the different instance which may be empty.

https://github.com/mitsuba-renderer/drjit/blob/07e9da811e7284b87fa292472b30ec4465592eef/src/autodiff/autodiff.cpp#L398


Maybe Clang11 and Clang14 have different default compile parameters, and Clang14 will staticly link the global variable, which will cause the global variable constructed more than once.

**Comment by anewusername77** [comment]
> ## Summary
> I wanted to optimize the vertex positions of a mesh but encountered some strange error.
> 
> ## System configuration
> * Platform: Ubuntu 22.04
> * Compiler: clang 14
> * Python version: 3.9
> * Mitsuba 3 version: 3.0.1
> * Compiled variants:
>   
>   * `scalar_rgb`
>   * `cuda_ad_rgb`
> 
> ## Description
> I called `dr.enable_grad()` on the vertex positions and then `params.update()` to update the scene, and Dr. JIT complained that an unknown variable was referenced. But the error did not occur when the parameter was vertex normals.
> 
> ## Steps to reproduce
> ```
> import drjit as dr
> import mitsuba as mi
> 
> mi.set_variant('cuda_ad_rgb')
> 
> key = 'redwall.vertex_positions'
> 
> # Loading the Cornell Box scene shipped along with Mitsuba 3
> scene = mi.load_file('tutorials/scenes/cbox.xml', res=128, integrator='prb_reparam')
> 
> params = mi.traverse(scene)
> 
> opt = mi.ad.Adam(lr=0.02)
> opt[key] = params[key]
> params.update(opt)
> ```
> 
> ```
> Critical failure in Dr.Jit AD backend: referenced an unknown variable a2494425056!
> Aborted (core dumped)
> ```

is it possible to use torch.optim to optimize params in mitsuba?

**Comment by njroussel** [comment]
Yes, take a look at this tutorial: https://mitsuba.readthedocs.io/en/latest/src/inverse_rendering/pytorch_mitsuba_interoperability.html

Again, please open a new issue/discussion for such questions. This is barely related to the original topic.

**Comment by anewusername77** [comment]
> Yes, take a look at this tutorial: https://mitsuba.readthedocs.io/en/latest/src/inverse_rendering/pytorch_mitsuba_interoperability.html
> 
> Again, please open a new issue/discussion for such questions. This is barely related to the original topic.

new issue is here: https://github.com/mitsuba-renderer/mitsuba3/issues/459
thanks a lot!
